Upgrade ASM to 7.1

Godin opened the pull request:
Fixes #839

Godin commented:
@marchof this is because you picked PR which is still in "implementation" 😆 btw hard to predict url for changelog before creation of PR 😉

marchof commented on org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java:
```diff
@@ -158,18 +158,16 @@ private InstrSupport() {
-		return (short) (((b[MAJOR_VERSION_INDEX] & 0xFF) << 8)
-				| (b[MAJOR_VERSION_INDEX + 1] & 0xFF));
+	public static int getVersionMajor(final ClassReader reader) {
+		final int firstConstantPoolEntryOffset = reader.getItem(1) - 1;
```
Why are you going relative to the contant pool entry? Isn't it always at the same absolute position? I'm asking because we have another method `classReaderFor()` in the same class which uses a fixed lookup index. Should be consistent.

Godin replied:
@marchof recall #850 - there is constructor [`ClassReader(byte[] classFileBuffer, int classFileOffset, int classFileLength)`](https://gitlab.ow2.org/asm/asm/blob/ASM_7_1/asm/src/main/java/org/objectweb/asm/ClassReader.java#L165-177).

Godin replied:
@marchof so relative to the constant pool entry seems safer to me, even if currently our code does not pass such instances of `ClassReader` into this method.

Godin replied:
I can update unit test to show/cover this case.

Or we can create subclass of `ClassReader` that will returned by `classReaderFor` and required here.

marchof replied:
@Godin ASM API is really wired here: it has public `readX()` methods but keeps the offset private. My proposals to resolve this:

* Use absolute positions as the offset constructor cannot be used any more due to #850 or
* Add a inline comment to explain why it is done this way

marchof commented on org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java:
```diff
@@ -238,7 +236,7 @@ public static void push(final MethodVisitor mv, final int value) {
 	public static ClassReader classReaderFor(final byte[] b) {
 		final byte[] originalVersion = new byte[] { b[4], b[5], b[6], b[7] };
-		if (getVersionMajor(b) == Opcodes.V12 + 1) {
+		if (b[6] == 0x00 && b[7] == 0x39) {
```
Magic number? The intend was more clear with `Opcodes.V12 + 1`.

Godin replied:
@marchof inlining of former implementation of `getVersionMajor` into this method results in

```java
final int majorVersion = (short) (((b[MAJOR_VERSION_INDEX] & 0xFF) << 8)
		| (b[MAJOR_VERSION_INDEX + 1] & 0xFF));

if (majorVersion == Opcodes.V12 + 1)
```

which IMO contains far more magic and far less readable than

```
if (b[6] == 0x00 && b[7] == 0x39)
```

Can also add comment.

Readable and non-magical alternative

```
final int majorVersion = java.nio.ByteBuffer.wrap(b)
		.order(java.nio.ByteOrder.BIG_ENDIAN).getShort(6);
```

looks like a bit of overkill.

Godin replied:
BTW numbers `6`, `7` and `0x39` are even shown in https://en.wikipedia.org/wiki/Java_class_file 😉

marchof replied:
So why inlining? If we keep the previous method (in addition to the new Reader version) we have testable and readable code ;)

marchof commented:
@Godin If you like we can merge as is and check in a separate PR how we can streamline version reading/writing. Ok?

marchof commented:
@Godin What about requesting a new public API `ClassReader.getVersion()` from the ASM project?

marchof commented:
@Godin With explicit methods to read versions the code becomes way more understandable:

	public static ClassReader classReaderFor(final byte[] b) {
		final int originalVersion = getVersion(b);
		if (getVersionMajor(b) == Opcodes.V12 + 1) {
			// temporary downgrade version for ASM
			setVersion(Opcodes.V12, b);
		}
		final ClassReader classReader = new ClassReader(b);
		setVersion(originalVersion, b);
		return classReader;
	}

Godin commented:
@marchof 

> What about requesting a new public API ClassReader.getVersion() from the ASM project?

Feel free to ask them :wink: :innocent:

----

Speaking about understandability, testability and magic :crystal_ball: - actually implementation of `getVersionMajor(byte[])` in `0.8.3`

https://github.com/jacoco/jacoco/blob/6cd3f0bd8e348f8fba7bffec5225407151f1cc91/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java#L170-L173

is incorrect, as well as

```
final int majorVersion = java.nio.ByteBuffer.wrap(b)
		.order(java.nio.ByteOrder.BIG_ENDIAN).getShort(6);
```

because [according to JVMS](https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.1) `major_version` is **unsigned** :wink:

For example let's consider `major_version` that with current temps (2 per year) will be approximately in 16393 years from now :smiling_imp: - `32786` (`0x8012`) and thus following test

```
final byte[] bytes = new byte[] { //
		(byte) 0xCA, (byte) 0xFE, (byte) 0xBA, (byte) 0xBE, // magic
		0x00, 0x00, // minor_version
		(byte) 0x80, (byte) 0x12 // major_version
};
assertEquals(0x8000, InstrSupport.getVersionMajor(bytes));
```

which fails

```
Expected :32786
Actual   :-32750
```

IMO such mistake is impossible in case of proposed exact comparison of bytes

```
if (b[6] == 0x80 && b[7] == 0x12)
```

:wink:

However proposed implementation of `getVersionMajor(ClassReader)`

https://github.com/jacoco/jacoco/blob/c7e2176bef48044cbf6dfd425f84a2dd012128fa/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java#L168-L171

has the same mistake because of [`readShort` which reads signed value](https://gitlab.ow2.org/asm/asm/blob/ASM_7_1/asm/src/main/java/org/objectweb/asm/ClassReader.java#L3366-3376).

Both come from the fact that [ASM itself uses same signed `readShort` to read `major_version`](https://gitlab.ow2.org/asm/asm/blob/ASM_7_1/asm/src/main/java/org/objectweb/asm/ClassReader.java#L193-196) :laughing: :rofl: This also means that instead of "downgrade" trick, we can actually do "upgrade" trick :rofl: :rofl:

----

Moreover `major_version` is **2 bytes** and so implementation of `needsFrames(int)`

https://github.com/jacoco/jacoco/blob/6cd3f0bd8e348f8fba7bffec5225407151f1cc91/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java#L182-L185

is also incorrect and will fail / require change earlier - in about 122 years :smiling_imp:

```
assertEquals(true, InstrSupport.needsFrames(0x0100));
```

Morever similarly to https://github.com/jacoco/jacoco/pull/851#discussion_r262829574 about `getVersionMajor(ClassReader)`, `needsFrames` currently always receives `major_version` and so doesn't need to "consider 1.1 anomaly" ? :wink:

----

I added correct implementations of `getMajorVersion(byte[])` and `setMajorVersion(int, byte[])`,
fixed, renamed and commented `getMajorVersion(ClassReader)`. And let's fix `needsFrames` separately.

marchof commented:
@Godin Thanks for fixing this right away!

> will fail / require change earlier - in about 122 years

Ok, we document this as know limitation -- along with > 32 dimensional arrays 🤣

marchof commented on org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java:
```diff
@@ -158,18 +158,50 @@ private InstrSupport() {
+	 *            reader to get information about the class
+	 * @return major version of bytecode
+	 * @see ClassReader#ClassReader(byte[], int, int)
+	 * @see #getMajorVersion(byte[])
```
This give me a warning because public API links to package private API (which is not part of JavaDoc). Remove link or make methods public?

Godin replied:
Done

marchof replied:
@Godin Now as it is public we have a new warning as parameter `b` is not documented 🙈

Godin replied:
Fixed
